### PR TITLE
[@types/relay-runtime]: Fix version 10.1.x

### PR DESCRIPTION
**Pull request by morrys**
Some fixes I found while developing react-relay-offline 3.0.0:

[RelayStoreTypes.js react flight](https://github.com/facebook/relay/blob/v10.1.2/packages/relay-runtime/store/RelayStoreTypes.js#L975-L992)
[RelayNetworkTypes.js react flight](https://github.com/facebook/relay/blob/v10.1.2/packages/relay-runtime/network/RelayNetworkTypes.js#L128-L145)
[RelayModernQueryExecutor](https://github.com/facebook/relay/blob/v10.1.2/packages/relay-runtime/store/RelayModernQueryExecutor.js#L68-L86)
[RelayRecordState.js](https://github.com/facebook/relay/blob/v10.1.2/packages/relay-runtime/store/RelayRecordState.js)
[ReaderNode.js](https://github.com/facebook/relay/blob/v10.1.2/packages/relay-runtime/util/ReaderNode.js)
[NormalizationNode.js](https://github.com/facebook/relay/blob/v10.1.2/packages/relay-runtime/util/NormalizationNode.js)
[index.js](https://github.com/facebook/relay/blob/v10.1.2/packages/relay-runtime/index.js)

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [X] Provide a URL to documentation or source code which provides context for the suggested changes: 

- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## relay-runtime ([<kbd>unpkg</kbd>](https://unpkg.com/browse/relay-runtime@latest/))
was missing the following properties:
1. ReplaySubject
1. getNode
1. reportMissingRequiredFields
1. isFragment
1. isInlineDataFragment

<details>
<summary>as well as these 13 other properties...</summary>
<p>isRequest, MutationTypes, RangeOperations, MutationHandlers, VIEWER_ID, VIEWER_TYPE, PreloadableQueryRegistry, createPayloadFor3DField, RelayError, DEFAULT_HANDLE_KEY, isScalarAndEqual, recycleNodesInto, stableCopy</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 73799c658a77dc26c03ca174cfa21dd3cb68ab7a
</p>

File: types/relay-runtime/lib/network/RelayNetworkTypes.d.ts [hunk 1]
```diff
@@ -111,3 +111,25 @@ export type Uploadable = File | Blob;
 export interface UploadableMap {
     [key: string]: Uploadable;
 }
+
+/**
+ * React Flight tree created on the server.
+ */
+export type ReactFlightServerTree = any;
```
Comment (Stephen2) [inline types/relay-runtime/lib/network/RelayNetworkTypes.d.ts:118]: I think `unknown` is preferable to `any`, when translating from Flow `mixed`

File: types/relay-runtime/lib/network/RelayNetworkTypes.d.ts [hunk 2]
```diff
@@ -111,3 +111,25 @@ export type Uploadable = File | Blob;
...
+export type ReactFlightPayloadQuery = {
+    id: any;
+    module: any;
```
Comment (Stephen2) [inline types/relay-runtime/lib/network/RelayNetworkTypes.d.ts]: maybe `unknown` over `any`?

File: types/relay-runtime/lib/network/RelayNetworkTypes.d.ts [hunk 3]
```diff
@@ -111,3 +111,25 @@ export type Uploadable = File | Blob;
 export interface UploadableMap {
     [key: string]: Uploadable;
 }
+
+/**
+ * React Flight tree created on the server.
+ */
+export type ReactFlightServerTree = any;
+export type ReactFlightPayloadQuery = {
```
Comment (Stephen2) [inline types/relay-runtime/lib/network/RelayNetworkTypes.d.ts]: Looks like Flow marks all these properties as `readonly`.  Do we wanna do that for TypeScript types too?

File: types/relay-runtime/lib/network/RelayNetworkTypes.d.ts [hunk 4]
```diff
@@ -111,3 +111,25 @@ export type Uploadable = File | Blob;
...
+ * Data that is returned by a Flight compliant GraphQL server.
+ *
+ * - tree: an array of values that will be iterated and fed into
+ *     ReactFlightDOMRelayClient.
+ * - queries: an array of queries that the server preloaded for the client.
+ */
+export type ReactFlightPayloadData = {
+    tree: Array<ReactFlightServerTree>;
+    queries: Array<ReactFlightPayloadQuery>;
+};
```
Comment (Stephen2) [inline types/relay-runtime/lib/network/RelayNetworkTypes.d.ts]: `readonly` on properties to match Flow

File: types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts [hunk 5]
```diff
@@ -4,26 +4,41 @@ import {
...
 } from './RelayStoreTypes';
 import { GraphQLResponse } from '../network/RelayNetworkTypes';
 import { Sink, RelayObservable } from '../network/RelayObservable';
 import { GetDataID } from './RelayResponseNormalizer';
 
+export type ActiveState = 'active' | 'inactive';
+
 export interface ExecuteConfig {
     getDataID: GetDataID;
+    treatMissingFieldsAsNull: boolean;
```
Comment (Stephen2) [inline types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts]: maybe `readonly` (I'll stop saying this).  But anyway, I see the other properties in this interface aren't marked `readonly` so maybe we don't do that.

File: types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts [hunk 6]
```diff
@@ -4,26 +4,41 @@ import {
...
-    operationLoader: OperationLoader;
-    operationTracker?: OperationTracker;
-    optimisticConfig: OptimisticResponseConfig;
+    operationExecutions: Map<string, ActiveState>;
+    operationLoader: OperationLoader | null | undefined;
+    operationTracker?: OperationTracker | null | undefined;
+    optimisticConfig: OptimisticResponseConfig | null | undefined;
     publishQueue: PublishQueue;
-    scheduler?: TaskScheduler;
+    reactFlightPayloadDeserializer?: ReactFlightPayloadDeserializer | null;
```
Comment (Stephen2) [inline types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts]: missed `| undefined`.  I know the `?` means `undefined` is allowed, but then we should be consistent with how `operationTrcker?` is written above

File: types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts [hunk 7]
```diff
@@ -4,26 +4,41 @@ import {
...
+    scheduler?: TaskScheduler | null;
```
Comment (Stephen2) [inline types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts]: `| undefined` (or be consistent throughout)

File: types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts [hunk 8]
```diff
@@ -4,26 +4,41 @@ import {
...
     sink: Sink<GraphQLResponse>;
     source: RelayObservable<GraphQLResponse>;
-    updater?: SelectorStoreUpdater;
+    store: Store;
+    updater?: SelectorStoreUpdater | null;
```
Comment (Stephen2) [inline types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts]: `| undefined`

File: types/relay-runtime/lib/store/RelayStoreTypes.d.ts [hunk 9]
```diff
@@ -907,3 +914,16 @@ export interface PublishQueue {
      */
     run(): ReadonlyArray<RequestDescriptor>;
 }
+
+/**
+ * ReactFlightDOMRelayClient processes a ReactFlightServerTree into a
+ * ReactFlightClientResponse object. readRoot() can suspend.
+ */
+export type ReactFlightClientResponse = { readRoot: () => any };
```
Comment (Stephen2) [inline types/relay-runtime/lib/store/RelayStoreTypes.d.ts]: `unknown` over `any` (I think)

File: types/relay-runtime/lib/store/RelayStoreTypes.d.ts [hunk 10]
```diff
@@ -907,3 +914,16 @@ export interface PublishQueue {
...
+
+export type ReactFlightReachableQuery = {
+    module: any;
```
Comment (Stephen2) [inline types/relay-runtime/lib/store/RelayStoreTypes.d.ts]: `module: mixed`

File: types/relay-runtime/lib/util/NormalizationNode.d.ts [hunk 11]
```diff
@@ -55,6 +41,14 @@ export interface NormalizationScalarField {
     readonly storageKey: string | null | undefined;
 }
 
+export type NormalizationFlightField = {
+    readonly kind: string; // 'FlightField';
```
Comment (Stephen2) [inline types/relay-runtime/lib/util/NormalizationNode.d.ts:45]: why `string` over `FlightField` here?

File: types/relay-runtime/lib/util/ReaderNode.d.ts [hunk 12]
```diff
@@ -82,13 +68,45 @@ export interface ReaderScalarField {
     readonly storageKey: string | null | undefined;
 }
 
+export type ReaderFlightField = {
+    readonly kind: string; // 'FlightField',
```
Comment (Stephen2) [inline types/relay-runtime/lib/util/ReaderNode.d.ts:72]: and here

File: types/relay-runtime/lib/util/ReaderNode.d.ts [hunk 13]
```diff
@@ -82,13 +68,45 @@ export interface ReaderScalarField {
...
+    readonly alias: string | null | undefined;
+    readonly name: string;
+    readonly args: ReadonlyArray<ReaderArgument> | null | undefined;
+    readonly storageKey: string | null | undefined;
+};
+
+export type ReaderDefer = {
+    readonly kind: string; // 'Defer',
```
Comment (Stephen2) [inline types/relay-runtime/lib/util/ReaderNode.d.ts:80]: same

File: types/relay-runtime/lib/util/ReaderNode.d.ts [hunk 14]
```diff
@@ -82,13 +68,45 @@ export interface ReaderScalarField {
...
+    readonly selections: ReadonlyArray<ReaderSelection>;
+};
+
+export type ReaderStream = {
+    readonly kind: string; // 'Stream',
```
Comment (Stephen2) [inline types/relay-runtime/lib/util/ReaderNode.d.ts:85]: same

File: types/relay-runtime/lib/util/ReaderNode.d.ts [hunk 15]
```diff
@@ -82,13 +68,45 @@ export interface ReaderScalarField {
...
+    readonly selections: ReadonlyArray<ReaderSelection>;
+};
+
+export type RequiredFieldAction = 'NONE' | 'LOG' | 'THROW';
+
+export type ReaderRequiredField = {
+    readonly kind: string; // 'RequiredField'
```
Comment (Stephen2) [inline types/relay-runtime/lib/util/ReaderNode.d.ts:92]: same

File: types/relay-runtime/lib/network/RelayNetworkTypes.d.ts (hunk 3 above)
Comment (Stephen2) [reply types/relay-runtime/lib/network/RelayNetworkTypes.d.ts]: oh, done in later commit :+1:

File: types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts [hunk 16]
```diff
@@ -20,7 +20,7 @@ export interface ExecuteConfig {
     readonly operation: OperationDescriptor;
     readonly operationExecutions: Map<string, ActiveState>;
     readonly operationLoader: OperationLoader | null | undefined;
-    readonly operationTracker?: OperationTracker | null | undefined;
+    readonly operationTracker?: OperationTracker | null;
```
Comment (Stephen2) [inline types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts:23]: I don't understand why `undefined` was removed, though I see the `?` covers it as an option

**Review by Stephen2 — COMMENTED** [review]
I'll leave for someone else to make final call on some of these `unknown > any` and string constants.  But I hope that helped for a first review :+1: 

This was super good PR

File: types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts (hunk 6 above)
Comment (Stephen2) [reply types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts]: I think you made `operationTracker` consistent in later commits

File: types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts (hunk 7 above)
Comment (Stephen2) [reply types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts]: I think you made operationTracker consistent in later commits

File: types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts (hunk 8 above)
Comment (Stephen2) [reply types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts]: Probably not, you're consistently not doing this now.

File: types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts (hunk 16 above)
Comment (Stephen2) [reply types/relay-runtime/lib/store/RelayModernQueryExecutor.d.ts:23]: OK, I can see you're just being consistent now.  Maybe even a linter picked this up.  EIther was :+1:

**Comment by morrys** [comment]
Thanks @stephen2,
this weekend I'm away from the pc :) Monday I look at the suggestions;)

**Comment by morrys** [comment]
I used string because the others had it 👍 
let's see what @alloy thinks

**Comment by morrys** [comment]
@alloy friendly ping, to tell you that this PR is useful for me to continue the evolution of react-relay-offline :)

If you are on vacation, watch it later 👍

**Review by Stephen2 — APPROVED** [review]
None of my comments are blocking, this can merge and be reviewed after merge if desired

File: types/relay-runtime/lib/util/NormalizationNode.d.ts (hunk 11 above)
Comment (maraisr) [reply types/relay-runtime/lib/util/NormalizationNode.d.ts:45]: these "kind" fields break things.. The general gist here is to keep them as string.

**Review by maraisr — APPROVED** [review]
LGTM! I think one thing we need here is tests, but will merge if you can loop back at some point and add tests!

**Comment by maraisr** [comment]
Ready to merge

**Comment by alloy** [comment]
> @alloy friendly ping, to tell you that this PR is useful for me to continue the evolution of react-relay-offline :)
> 
> If you are on vacation, watch it later 👍

I wish, in reality I’ve been out sick 🤒 

Anyways, indeed Flow’s `mixed` should be converted to `unknown`.

**Pull request merged**